### ISSUE TITLE
"Helper not found while proving induction lemma" with functional extensionality unset

### ISSUE THREAD

**Issue by MrSet** [mattam82/Coq-Equations#306]
When trying to use Equations to define a non-structurally recursive function with functional extensionality I get the error:

```Error: Anomaly "Helper not found while proving induction lemma."```

The minimal example I could find is:
```coq
From mathcomp Require Import
     all_ssreflect.

From Equations Require Import Equations.
Unset Equations With Funext.

Equations test (s : seq bool)  : seq bool by wf (size s) lt:=
  test pn with true => {
  | true := [::];
  | false := [::]  }. 
```

With functional extensionality set, it is defined without any problems.

**Comment by mattam82** [comment]
Thanks for the report! I will look into this.

**Comment by yforster** [comment]
I also have this problem (under `coq-equations.1.2.3+8.12`). For the record, a simplified version of my code is below.

I'm also using a `with` clause, and I also have `Funext` unset.

Did you make any progress on the issue?

```coq
Require Import List.

Definition dec (X: Prop) : Type := {X} + {~ X}.

From Equations Require Import Equations.
    
Definition dec_In {X} (x : X) L : dec (In x L).
Admitted.
  
Section fixes.
  
  Variables (X Y : Type).

  Variable f : X -> Y.
    
  Unset Equations With Funext.

  Equations γ (C : list (X * Y)) : X -> X by wf (length C) lt := 
    γ C x with dec_In (f x) (map snd C) => {
      | left H := x;
      | right _ := x
    }.
```

**Comment by mattam82** [comment]
Sorry for the delay, seems like a simple bug in the handling of the Funext option.

**Comment by mattam82** [comment]
A quick and dirty workaround is to redefine the problematic tactic like this:

```
Require Import Equations.Prop.Subterm Equations.Prop.DepElim.
From Equations Require Import Equations.

Unset Equations With Funext.

Parameter size : forall {A}, list A -> nat.

Set Equations Debug.
Ltac Subterm.unfold_FixWf ::=
match goal with
     |- context [ @FixWf ?A ?R ?WF ?P ?f ?x ] =>
     let step := fresh in
     set(step := fun y (_ : R y x) => @FixWf A R WF P f y) in *;
     unshelve erewrite (@FixWf_unfold_step A R WF P f x _ step);
     [red; intros; simp_sigmas (* Extensionality proof *)
     |hidebody step; DepElim.red_eq_lhs (* Unfold the functional *)
     |reflexivity]
end.

Equations test (s : list bool) : list bool by wf (size s) lt:=
  test pn with true => {
  | true := nil;
  | false := nil }.

  Next Obligation. Defined.
  Print test_elim.
```

That's rather fragile though, the ML code needs a fix. It basically applies the wrong tactic to the extensionality subgoal.

**Comment by yforster** [comment]
Thanks a lot, indeed in my use case I now get the unfold equation as proof goal.

Do you have any example code how to prove such an unfolding equations manually?

**Comment by mattam82** [comment]
Usually, you just need to do the same case split as your function does, and use reflexivity.

**Comment by mattam82** [comment]
It might be that the proper fix will actually solve the proof for you though, it used to use a broken proof script, and should now solve more goals :)

**Comment by yforster** [comment]
Ha, indeed :) I cherry-picked the fix back to 8.12 and it worked. Thanks a lot!
